### ISSUE TITLE
Kops --wait

### ISSUE THREAD

**Issue by ajohnstone** [kubernetes/kops#1517]
Executing both ```kops cluster create ...``` and ```kops cluster update``` took 2.57 minutes for a new cluster. However the kubernetes cluster is not ready for use at this point.

E.g.
```
$ kubectl get pods
Unable to connect to the server: EOF

$ kops-1.5.0-alpha3 validate  cluster
Using cluster from kubectl context: xxxxxx.photobox.com

Validating cluster xxxxxx.photobox.com


Cannot get nodes for "xxxxxx.photobox.com": Get https://api.xxxxxx.photobox.com/api/v1/nodes: EOF

$ kops-1.5.0-alpha3 validate  cluster

Using cluster from kubectl context: xxxxxx.photobox.com

Validating cluster xxxxxx.photobox.com


Cannot get nodes for "xxxxxx.photobox.com": Get https://api.xxxxxx.photobox.com/api/v1/nodes: EOF

$ kops-1.5.0-alpha3 validate  cluster
Using cluster from kubectl context: xxxxxx.photobox.com

Validating cluster xxxxxx.photobox.com

INSTANCE GROUPS
NAME			ROLE	MACHINETYPE	MIN	MAX	SUBNETS
bastions		Bastion	t2.micro	1	1	utility-eu-west-1b,utility-eu-west-1c,utility-eu-west-1a
master-eu-west-1a	Master	m3.large	1	1	eu-west-1a
master-eu-west-1b	Master	m3.large	1	1	eu-west-1b
master-eu-west-1c	Master	m3.large	1	1	eu-west-1c
nodes			Node	m3.large	5	5	eu-west-1b,eu-west-1c,eu-west-1a

NODE STATUS
NAME						ROLE	READY
ip-10-0-102-114.eu-west-1.compute.internal	master	True
ip-10-0-104-134.eu-west-1.compute.internal	node	False
ip-10-0-120-191.eu-west-1.compute.internal	node	True
ip-10-0-47-148.eu-west-1.compute.internal	node	False
ip-10-0-58-87.eu-west-1.compute.internal	master	True
ip-10-0-78-49.eu-west-1.compute.internal	node	False
ip-10-0-80-192.eu-west-1.compute.internal	master	True
ip-10-0-84-31.eu-west-1.compute.internal	node	False

Validation Failed
Master(s) Not Ready 0 out of 3.
Node(s) Not Ready   4 out of 5.

Your cluster xxxxxx.photobox.com is NOT ready.
```

It would be ideal to have kops with a --wait option for the cluster to be ready.
There is a similar ticket https://github.com/kubernetes/kops/issues/139.

There are a few example scripts to wait for a cluster to be initialised, however would be ideal to have this part of kops.

Also to note @chrislovecnm kops cluster validate doesn't do anything with ```kubectl get cs``` maybe ideal to add into the kops validate cluster?

```
kubectl config use-context ${CLUSTER_NAME}
echo -n "Waiting for cluster components to become ready."
until [ $(kubectl get cs 2> /dev/null| grep -e Healthy | wc -l | xargs) -ge 4 ]
do
  echo -n "."
  sleep 1
done
echo "ok"

echo -n "Waiting for minimum nodes to become ready."
min_nodes=$(kops get ig nodes --name ${CLUSTER_NAME} | grep nodes | awk '{print $4}')
until [ "$(kubectl get nodes 2> /dev/null| grep -v master | grep -e Ready | wc -l | xargs)" == "$min_nodes" ]
do
  echo -n "."
  sleep 1
done
echo "ok"
```
@shadoi

or

```
while [ 1 ]; do
    kops validate cluster && break || sleep 30
done;
```

**Comment by chrislovecnm** [comment]
I am not a huge fan of putting in a wait.  But will let @kris-nova and @justinsb weigh in as well.

@ajohnstone adding issue for adding kubectl get cs equivalent in kops validate

**Comment by justinsb** [comment]
What is the objection to a wait command @chrislovecnm ?

**Comment by kris-nova** [comment]
I am all for the `--wait` flag! 

### Concerns/Questions

1. Can we do any better than `--wait`? Maybe `--validate`?

2 .My other concern would be a line from the story here:

> It would be ideal to have kops with a --wait option for the cluster to be ready.

With of course the magic phrase being `cluster to be ready`.. As long as we can agree on what that means (and probably more importantly what it doesn't mean) and clearly stick to that I am good. 

Should we use `kops validate` logic here? 
Verify the API is listening? 
Healthz? 
etc..

3. Validation Timeout

So I think we have 2 spans of time to track. This would of course be the time kops would wait from some **instant** in time - until we finally give up and error.

4. Minimum Valid Span

So the other span we would need to track would be the time from the **instant** we receive a valid cluster until some arbitrary span has elapsed (while still receiving a valid message)



So basically - it seems simple to have kops hang.. but we would need to clarify a few things, and make clear expectations around what kops will and wont offer/promise the user.

**Comment by chrislovecnm** [comment]
I like to keep this async or a single call.  Implementing something I can do with a while loop in bash, just seems overkill.  But that is just me. :man_shrugging: 

I can see that there is value to the user.  I like having the validate wait, with a configurable amount of loops, more palatable.  Putting a wait into the create, just seems overkill.  But eh ... what do our users need :)

**Comment by kenden** [comment]
@chrislovecnm I have that `while` loop in bash inside a makefile:

```
.PHONY: wait_for_cluster_ready
wait_for_cluster_ready:
  @max_wait=900; \
  while [[ $${max_wait} -gt 0 ]]; do \
    ${KOPS} validate cluster ${CLUSTER_NAME} --state ${KOPS_STATE} && break || sleep 30; \
    max_wait=$$((max_wait - 30)); \
    echo "Waited 30 seconds. Still waiting max. $${max_wait}"; \
  done; \
  if [[ $${max_wait} -le 0 ]]; then \
    echo "Timeout: cluster does not validate after 15 minutes"; \
    exit 1; \
  fi
```
It's a bit "urgh".
I think it would look much cleaner with just a `--wait` on the `create/update` commands.
There are similar tickets for helm to make the user's life easier BTW.
https://github.com/kubernetes/helm/issues/2114
https://github.com/kubernetes/helm/issues/1805

**Comment by techdragon** [comment]
@ajohnstone do you want to reopen this?

**Comment by jbialy** [comment]
Could this be reopened once again? I think it'd be a very useful feature to have!

**Comment by bryan-rhm** [comment]
I'm waiting for this feature too, this should be reopened

**Comment by TomBloo** [comment]
Re-open!

**Comment by SCLogo** [comment]
no updates on this ?

**Comment by agilgur5** [comment]
Would adding `--wait` to `kops validate cluster` be more acceptable than adding to `kops create cluster`? i.e. `kops validate cluster --wait` would loop until the cluster passes validation.

Right now I've just added this script/function to do so:
```bash
kops_validate_loop() {
  kops_exit_code=1

  # `kops validate cluster` should exit with code 0 to be successful
  while [[ $kops_exit_code -ne 0 ]]; do
    # subshell to workaround set -e
    kops_exit_code=$(kops validate cluster \
      --name=$CONTEXT \
      --state=$KOPS_STATE_STORE \
      > /dev/null; echo $?)

    if [[ $kops_exit_code -ne 0 ]]; then
      # same as kops
      echo "Cluster did not pass validation, will try again in 30s..."
      sleep 30
    fi
  done
}
```
(note that `$CONTEXT` and `$KOPS_STATE_STORE` must be set beforehand if you're planning to use that)

**Comment by olemarkus** [comment]
Doesn't `kops validate cluster` have a `--wait` flag now?
` kops validate cluster --wait 60s`

**Comment by olemarkus** [comment]
`kops validate` alone isn't a valid command though. So the flags are only the global flag + the help flag. But I'll see if I can make the examples a bit more clear.

**Comment by agilgur5** [comment]
Thanks @olemarkus changes in #9333 would've definitely made me notice that. Yea I recognize it's not a valid command but it previously gave details about `validate cluster` so I either didn't realize there was a different doc or that the `validate cluster` doc had a tiny difference. Now they're a good bit different from each so I think that's easier to tell. This issue is still first on a google search but hopefully there's now a handful of ways to discover that flag and not miss it 🙂
